go/config/config.go

```diff
@@ -0,0 +1,184 @@
+// Package config implements global configuration options.
+package config
+
+import (
+	"fmt"
+
+	"github.com/a8m/envsubst"
+	"gopkg.in/yaml.v3"
+
+	tm "github.com/oasisprotocol/oasis-core/go/consensus/tendermint/config"
+	ias "github.com/oasisprotocol/oasis-core/go/ias/config"
+	common "github.com/oasisprotocol/oasis-core/go/oasis-node/cmd/common/config"
+	pprof "github.com/oasisprotocol/oasis-core/go/oasis-node/cmd/common/pprof/config"
+	p2p "github.com/oasisprotocol/oasis-core/go/p2p/config"
+	runtimeRegistry "github.com/oasisprotocol/oasis-core/go/runtime/registry/config"
+	workerCommon "github.com/oasisprotocol/oasis-core/go/worker/common/config"
+	workerKM "github.com/oasisprotocol/oasis-core/go/worker/keymanager/config"
+	workerRegistration "github.com/oasisprotocol/oasis-core/go/worker/registration/config"
+	workerSentry "github.com/oasisprotocol/oasis-core/go/worker/sentry/config"
+	workerStorage "github.com/oasisprotocol/oasis-core/go/worker/storage/config"
+)
+
+// NodeMode is the node mode.
+type NodeMode string
+
+const (
+	// ModeValidator is the name of the validator node mode.
+	ModeValidator NodeMode = "validator"
+	// ModeCompute is the name of the compute node mode.
+	ModeCompute NodeMode = "compute"
+	// ModeKeyManager is the name of the key manager node mode.
+	ModeKeyManager NodeMode = "keymanager"
+	// ModeClient is the name of the client node mode.
+	ModeClient NodeMode = "client"
+	// ModeStatelessClient is the name of the stateless client node mode.
+	ModeStatelessClient NodeMode = "client-stateless"
+	// ModeSeed is the name of the seed node mode.
+	ModeSeed NodeMode = "seed"
+	// ModeArchive is the name of the archive node mode.
+	ModeArchive NodeMode = "archive"
+)
+
+// IsClientOnly returns true iff the mode is one that has the node running
+// as a client for all configured runtimes.
+func (m NodeMode) IsClientOnly() bool {
+	switch m {
+	case ModeClient, ModeStatelessClient:
+		return true
+	}
+	return false
+}
+
+// HasLocalStorage returns true iff the mode is one that has local storage.
+func (m NodeMode) HasLocalStorage() bool {
+	switch m {
+	case ModeClient, ModeCompute:
+		return true
+	}
+	return false
+}
+
+// GlobalConfig holds the global configuration options.
+var GlobalConfig Config
+
+// Config is the top-level configuration structure.
+type Config struct {
+	// Oasis node mode (validator, non-validator, compute, keymanager, etc.).
+	Mode NodeMode `yaml:"mode"`
+
+	// Node is a consensus validator.
+	// This additional option exists because it is currently possible to
+	// run a node that is simultaneously a validator and a compute node.
+	ConsensusValidator bool `yaml:"consensus_validator"`
```

Comment:
Do we anticipate other options under the `consensus` hierarchy in the future?  I flattened this because I think it's ugly to have just a single option in a struct (the option itself might as well have been named just `validator`).
I can add an extra struct just for this, but since this is a nice opportunity to make the config options nicer, maybe we can keep it as it is now?

Reply:
See my other comment, I think we should put everything under `consensus` and drop the `tendermint` thing.